determined-ai/determined
# refactor: consolidate experiment list selection state

## Description

This refactors some items related to the experiment list selection ahead of fixing the bulk action apis:

1) selection state (selected/excluded experiment ids, selectAll state) is consolidated into a single object. Previously, selected experiment ids, excluded experiment ids, and the selectall state were stored separately. This is problematic -- because the selected experiment ids were the source of truth for which experiments were selected, the stored list would grow as the user paged through the results when in selectall mode. we now determine the final list of selected elements on the fly.

2) grid selection is now fully determined from the id selection. previously, the grid selection and id selection were maintained separately. this required a lot of code ensuring they stayed in sync. we now only derive the grid selection from the selected id list.

3) handleSelectionChange now no longer requires row ranges for `add-all` or `remove-all` change types. this allows for functions and effects that solely dealt with these change types to no longer rely on the experiments array.

4) project experiment list settings is no longer using useSettings. because the grid selection now only updates when the settings updates, the act of updating the selection felt slower than when using useSettings. we now directly interface with the settings store at the project level -- global settings still use useSettings.

5) new hook: useTypedParams. In order to keep the functionality of the `compare` parameter, we introduce a useTypedParams hook which allows typesafe parsing and setting of query params similar to the settings store api. this should allow us to finally fully migrate off of useSettings.


## Test Plan

Parameter/setttings behavior:
- navigate to a project experiment list with over 20 experiments in the list
- turn off infinite scroll and reload the page
- the experiment list should remain paginated
- click onto the second page
- the url should contain the query parameter `page=1`
- reload the page
- the experiment list should load the second page
- click back onto the first page
- the url should no longer contain the page query parameter
- toggle the comparison view on
- the url should contain the query parameter `compare=true`
- reload the page
- the experiment list should be in comparison mode
- remove the query parameter `compare=true` and reload the page
- the experiment list should remain in comparison mode
- add the query paramter `compare=false` and reload the page
- the experiment list should not be in comparison mode

Selection behavior:
- navigate to a project experiment list with over 20 experiments in the list
- select a single experiment in the list
- reload the page
- the experiment should remain selected
- navigate to the next page
- select a single experiment in the list
- navigate to the first page
- the selection status at the top of the table should show that one experiment is selected
- hold shift and select an experiment
- the experiment should be added to the selection
- in the selection column menu, choose select all
- the selection status at the top of the table should show that all experiments are selected
- deselect an experiment by clicking a checkbox
- the selection status at the top of the table should respond likewise



## Checklist

  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.

## Ticket
this is base work for WEB-1937
